File: kernel/src/main/kotlin/ninja/blacknet/rpc/v2/PeerInfo.kt
/*
 * Copyright (c) 2018-2024 Pavel Vasin
 *
 * Licensed under the Jelurida Public License version 1.1
 * for the Blacknet Public Blockchain Platform (the "License");
 * you may not use this file except in compliance with the License.
 * See the LICENSE.txt file at the top-level directory of this distribution.
 */

package ninja.blacknet.rpc.v2

import java.math.BigInteger
import kotlinx.serialization.Serializable
import ninja.blacknet.crypto.BigIntegerSerializer
import ninja.blacknet.crypto.Hash
import ninja.blacknet.db.CoinDB
import ninja.blacknet.db.Genesis
import ninja.blacknet.network.Connection
import ninja.blacknet.network.Node
import ninja.blacknet.network.packet.BlockAnnounce
import ninja.blacknet.serialization.LongSerializer

@Serializable
class PeerInfo(
        val peerId: Long,
        val remoteAddress: String,
        val localAddress: String,
        val timeOffset: Long,
        val ping: Long,
        val protocolVersion: Int,
        val agent: String,
        val outgoing: Boolean,
        val banScore: Int,
        @Serializable(with = LongSerializer::class)
        val feeFilter: Long,
        val connectedAt: Long,
        val lastChain: ChainInfo,
        val requestedBlocks: Boolean,
        val totalBytesRead: Long,
        val totalBytesWritten: Long,
) {
    @Serializable
    class ChainInfo(
            val chain: Hash,
            @Serializable(with = BigIntegerSerializer::class)
            val cumulativeDifficulty: BigInteger,
            val fork: Boolean
    ) {
        constructor(chain: BlockAnnounce, fork: Boolean) : this(
                chain.hash,
                chain.cumulativeDifficulty,
                fork
        )

        companion object {
            fun get(chain: BlockAnnounce, forkCache: HashMap<Hash, Boolean>): ChainInfo {
                val fork = forkCache.computeIfAbsent(chain.hash) { !CoinDB.blockIndexes.contains(it.bytes) }
                return ChainInfo(chain, fork)
            }
        }
    }

    companion object {
        fun get(connection: Connection, forkCache: HashMap<Hash, Boolean>): PeerInfo {
            return PeerInfo(
                    connection.peerId,
                    connection.remoteAddress.toString(),
                    connection.localAddress.toString(),
                    connection.timeOffset,
                    connection.ping,
                    connection.version,
                    connection.agent,
                    connection.state.isOutgoing(),
                    connection.dosScore(),
                    connection.feeFilter,
                    connection.connectedAt,
                    ChainInfo.get(connection.lastBlock, forkCache),
                    connection.requestedBlocks,
                    connection.getTotalBytesRead(),
                    connection.getTotalBytesWritten(),
            )
        }

        fun getAll(): List<PeerInfo> {
            val forkCache = HashMap<Hash, Boolean>()
            forkCache.put(Genesis.BLOCK_HASH, false)
            return Node.connections.map { PeerInfo.get(it, forkCache) }
        }
    }
}

File: json-rpc/src/main/rust/v2/peerinfo.rs
/*
 * Copyright (c) 2018-2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use crate::v2::{AmountInfo, EndpointInfo, HashInfo};
use blacknet_network::connection::Connection;
use blacknet_network::packet::BlockAnnounce;
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize)]
pub struct PeerInfo {
    peerId: u64,
    remoteAddress: EndpointInfo,
    localAddress: EndpointInfo,
    timeOffset: i64,
    ping: i64,
    protocolVersion: u32,
    agent: String,
    outgoing: bool,
    banScore: u8,
    feeFilter: AmountInfo,
    connectedAt: i64,
    lastChain: ChainInfo,
    requestedBlocks: bool,
    totalBytesRead: u64,
    totalBytesWritten: u64,
}

impl From<&Connection> for PeerInfo {
    fn from(connection: &Connection) -> Self {
        Self {
            peerId: connection.id(),
            remoteAddress: connection.remote_endpoint().into(),
            localAddress: connection.local_endpoint().into(),
            timeOffset: connection.time_offset().into(),
            ping: connection.ping().into(),
            protocolVersion: connection.version(),
            agent: connection.agent().into(),
            outgoing: connection.state().is_outgoing(),
            banScore: connection.dos_score(),
            feeFilter: connection.fee_filter().into(),
            connectedAt: connection.connected_at().into(),
            lastChain: connection.last_block().into(),
            requestedBlocks: connection.requested_blocks(),
            totalBytesRead: connection.total_bytes_read(),
            totalBytesWritten: connection.total_bytes_written(),
        }
    }
}

#[derive(Deserialize, Serialize)]
struct ChainInfo {
    chain: HashInfo,
    cumulativeDifficulty: String,
    fork: bool,
}

impl From<&BlockAnnounce> for ChainInfo {
    fn from(block_announce: &BlockAnnounce) -> Self {
        todo!();
    }
}
